### ISSUE TITLE
Multiple entrypoints into DICOM slides might confuse some software

### ISSUE THREAD

**Issue by bgilbert** [openslide/openslide#473]
### Operating system

any

### Platform

any

### OpenSlide version

4.0.0

### Slide format

DICOM

### Issue details

The DICOM driver, unlike other multi-file drivers, allows calling `openslide_open()` on any of the files comprising the slide.  This is convenient for command-line use, since it's not obvious by inspection which file is the "main" one.  However, it might break graphical software that assumes that every file where `openslide_detect_vendor(filename) != NULL` is a distinct slide.  For example, OpenSlide Python's `deepzoom_multiserver.py`:

<img src=https://github.com/openslide/openslide/assets/361374/bde3dc96-23c7-4b78-bc99-291c143c3ea9 height=400>

There is in fact a main image: the one with ImageType `{ORIGINAL,DERIVED}\PRIMARY\VOLUME\NONE`.  We could insist that that file should be the one specified to OpenSlide.  That would be nicer for GUIs, but would make command-line access to DICOM slides frustrating.  Graphical programs do have a workaround, though it's expensive: they could open each file and check its quickhash to detect duplicates.

There's also a third option: new API.  We could add an `openslide_detect_vendor_unique(filename)` or a `openslide_detect_vendor2(filename, OPENSLIDE_DETECT_UNIQUE)` that only returns non-NULL for the main file of a DICOM slide.

@jcupitt, what do you think?

**Comment by jcupitt** [comment]
I agree, this can be an issue in many contexts.

The API I've used successfully in another project is something like:

```C
char **openslide_get_files(openslide_t *osr);
```

It returns a list of the filenames in a slide, so:

```python
osr = openslide.open("3DHISTECH-1/000001.dcm")
osr.get_files() == ["3DHISTECH-1/000001.dcm", ""3DHISTECH-1/000002.dcm", ...]
```

It's faster and easier than testing files individually. It might possibly be useful for other slides that have multi-file formats, like MRXS.

**Comment by bgilbert** [comment]
Hmm, interesting approach.  That would still require applications to open each slide to see which files not to display, though.  That isn't great for affordances such as Open dialogs that want to filter files without opening any yet.

**Comment by jcupitt** [comment]
You only need to open each slide, not each file, so it's pretty quick.

It's handy for cases like "I want to move a slide to another directory, which files (and directories?) do I need to copy", which otherwise needs quite a bit of code.

**Comment by bgilbert** [comment]
Yeah, I can see how it could be useful.  Opening slides can require quite a bit of I/O, though.  For the Open dialog case, applications can currently use `openslide_detect_vendor()`, which is much more efficient.  That's why I was thinking a variant of that function would be useful.

**Comment by jcupitt** [comment]
Yes, I can see both being useful.
